modules/kafka/kafka-dest-driver.c

```diff
@@ -341,24 +487,39 @@ kafka_dd_init(LogPipe *s)
       return FALSE;
     }
 
-  self->kafka = _construct_client(self);
-  if (self->kafka == NULL)
+  if (!self->kafka)
     {
-      msg_error("kafka: error constructing kafka connection object, perhaps metadata.broker.list property is missing?",
-                evt_tag_str("topic", self->topic_name),
-                evt_tag_str("driver", self->super.super.super.id),
-                log_pipe_location_tag(&self->super.super.super.super));
-      return FALSE;
+      self->kafka = _construct_client(self);
+      if (self->kafka == NULL)
+        {
+          msg_error("kafka: error constructing kafka connection object, perhaps metadata.broker.list property is missing?",
+                    evt_tag_str("topic", self->topic_name->template),
+                    evt_tag_str("driver", self->super.super.super.id),
+                    log_pipe_location_tag(&self->super.super.super.super));
+          return FALSE;
+        }
     }
 
-  self->topic = _construct_topic(self);
-  if (self->topic == NULL)
+  if (_topic_name_is_a_template(self))
     {
-      msg_error("kafka: error constructing the kafka topic object",
-                evt_tag_str("topic", self->topic_name),
-                evt_tag_str("driver", self->super.super.super.id),
-                log_pipe_location_tag(&self->super.super.super.super));
-      return FALSE;
+      if (!_init_template_topic_name(self))
+        {
+          return FALSE;
+        }
+    }
+
+  else
+    {
+      self->topic = _construct_topic(self, self->topic_name->template);
+
```

Comment:
It would be great if you could extract the else branch as well, that would make reading this if easier, e.g.

```
if (is_topic_name_a_template(...) && !_init_template_topic_name(self) ||
   _init_literal_topic_name(self))
   return FALSE;
```

Reply:
forget the if example I gave above. That should be something like this instead:

if (_is_topic_name_a_template(self))
  {
    if (!_init_template_topic_name(self))
      return FALSE;
  }
else
  {
    if (!_init_literal_topic_name(self))
      return FALSE;
  }

Reply:
Ah, I think I found the perfect version, we need an _init_topic_name() function, that encapsulates the if, and that can be simpler as we don't need to continue with the rest of the kafka_dd_init() function.

```
static gboolean
_init_topic_name(KafkaDestDriver *self)
{
  if (_is_topic_name_a_template(self))
    return _init_template_topic_name(self);
  else
    return_init_literal_topic_name(self);
}

static gboolean
kafka_dd_init(...)
{
   ...
   if (!_init_topic_name(self))
    return FALSE;
   ...
}

```